### DISCUSSION TITLE
How to get atomic number and quantity of atoms in a molecule

### DISCUSSION THREAD

**Discussion by bertiewooster** [rdkit/rdkit#5339] [General] [answered]
Is there a built-in function to get the atomic number and quantity of atoms in a molecule? For example, for water [OH2], the desired result would be `{1: 2, 8: 1}` because hydrogen (atomic number 1) occurs 2 times, and oxygen (atomic number 8) occurs 1 time.

If not, my approach would be to build such a function by [looping through the atoms in the molecule](https://rdkit.readthedocs.io/en/latest/GettingStartedInPython.html#working-with-molecules)

```
>>> m = Chem.MolFromSmiles('C1OC1')
>>> for atom in m.GetAtoms():
...   print(atom.GetAtomicNum())
...
6
8
6
```

and accumulating their quantities in a dictionary I build.

**Comment by IchiruTake** [comment, marked as answer]
I think this is not intended. The idea of converting from Smiles to Mol is to constructing an appropriate graph out of it and making the atomic counter out of this is just an optional that not all RDKit users intended to. According to the algorithm, atomic counting is not a sub-product that could be extracted as an optional hyper-parameters. You can attempt to build your own function as the following.

``` python
m = Chem.MolFromSmiles(userInput)
atomic_count = defaultdict(lambda : 0)
for atom in m.GetAtoms():
     atomic_count[atom.GetAtomicNum()] += 1
print(atomic_count)

```

**Comment by bertiewooster** [comment]
Thanks! That function works as desired, except that hydrogen atoms aren't included in molecules by default (as I understand it), so I had to use .AddHs. I also wanted to include the charge in the dictionary. For future reference, here's my entire function:

```
def composition(molecule):
    """Get the composition of an RDKit molecule:
    Atomic counts, including hydrogen atoms, and any charge.
    For example, fluoride ion (chemical formula F-, SMILES string [F-])
    returns {9: 1, 0: -1}.

    :param molecule: The molecule to analyze
    :type some_input: An RDKit molecule
    :rtype: A dictionary.
    """
    # Check that there is a valid molecule
    if molecule:

        # Add hydrogen atoms--RDKit excludes them by default
        molecule_with_Hs = Chem.AddHs(molecule)
        comp = defaultdict(lambda: 0)

        # Get atom counts
        for atom in molecule_with_Hs.GetAtoms():
            comp[atom.GetAtomicNum()] += 1

        # If charged, add charge as "atomic number" 0
        charge = GetFormalCharge(molecule_with_Hs)
        if charge != 0:
            comp[0] = charge
        return comp
```
